### PR TITLE
Di 24015 add fields limitation validation

### PR DESCRIPTION
**Pull request by OdedAnswer**
there are some commits that shouldn't be there (namely the permissions) because it seems I mistakenly branched out from the wrong branch, but it's a branch that was already merged, so should be fine

### CODE REVIEW

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

File: client-side/src/app/collection/collection-list.component.ts [hunk 1]
```diff
@@ -78,6 +80,7 @@ export class CollectionListComponent implements OnInit {
         }).catch(error => {
             console.log(`could not get abstract schemes. error:${error}`);
         })
+        this.utilitiesService.getFieldLimit().then((value: number) => sessionStorage.setItem('fieldsLimit', value.toString()));
```
[the remark this answers is not in the document]
Comment (OdedAnswer) [reply client-side/src/app/collection/collection-list.component.ts:83]: when.. the session ends?

File: client-side/src/app/collection/form/tabs/general/general-tab.component.ts [hunk 2]
```diff
@@ -379,6 +374,16 @@ openFieldForm(name: string) {
     dialogConfig.data = new PepDialogData({
         content: FieldsFormComponent,
     })
+    if (dialogData.Mode == 'Add') {
+        if (Object.keys(this.collection.Fields).length >= this.fieldsLimit) {
```
Comment (roiaharon-pepperi) [inline client-side/src/app/collection/form/tabs/general/general-tab.component.ts:378]: it's not that easy to check it.. if the collection have contained resource type field you should include the inner count inside.
this is the logic on the server side
for (const fieldID of Object.keys(collectionObj.Fields!)) {
                if (collectionObj.Fields![fieldID].Type === 'ContainedResource') { // if field type is contained, count contained schema fields
                    const collectionFieldsCount = await this.getCollectionFieldsLength(collectionObj.Fields![fieldID].Resource!);
                    fieldsCount += collectionFieldsCount;
                } else {
                    fieldsCount++;
                }

File: server-side/api.ts [hunk 3]
```diff
@@ -453,6 +453,41 @@ export async function truncate(client: Client, request: Request) {
...
+    const collectionsService = new CollectionsService(client);
+    const utilities = new UtilitiesService(client);
+    const varRelationService: VarSettingsService = new VarSettingsService(utilities);
+    const apiService = new ApiService(client);
+    const documentsService = new DocumentsService(apiService, resourcesService);
+    switch (request.method) {
+        case 'GET': {
+            const resourcesForCollectionForm: ResourcesForCollectionForm = await resourcesService.getResourcesForCollectionForm(collectionName);
+            const collectionEvents = await collectionsService.getCollectionEvents(collectionName);
+            const fieldsLimit = await varRelationService.getSettingsByName(limitationTypes.Fields);
```
Comment (roiaharon-pepperi) [inline server-side/api.ts:468]: if the collection is of type 'contained' there is a different limit. you need to get a different key

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

File: client-side/src/app/collection/form/tabs/general/general-tab.component.ts [hunk 4]
```diff
@@ -388,6 +395,16 @@ openFieldForm(name: string) {
         if (value) {
             const fieldName = value.fieldName;
             this.collection.Fields[fieldName] = value.field;
+            if (this.getTotalFieldsCount(this.collection) > this.fieldsLimit) {
```
Comment (roiaharon-pepperi) [inline client-side/src/app/collection/form/tabs/general/general-tab.component.ts:398]: this check should be before you close the dialog. you can't delete the field, especially if it's "edit".

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

**Review by roiaharon-pepperi — APPROVED** [review]

**Pull request merged**